src/main/java/randoop/generation/GrtImpurity.java

```diff
@@ -0,0 +1,513 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import randoop.main.GenInputsAbstract;
+import randoop.main.RandoopBug;
+import randoop.operation.CallableOperation;
+import randoop.operation.ConstructorCall;
+import randoop.operation.MethodCall;
+import randoop.operation.TypedClassOperation;
+import randoop.operation.TypedOperation;
+import randoop.sequence.Sequence;
+import randoop.types.NonParameterizedType;
+import randoop.types.Type;
+import randoop.types.TypeTuple;
+import randoop.util.Randomness;
+
+/**
+ * Implements the "GRT Impurity" component, as described in "GRT: Program-Analysis-Guided Random
+ * Testing" by Ma et. al (ASE 2015): https://people.kth.se/~artho/papers/lei-ase2015.pdf.
+ *
+ * <p>The GRT Impurity component is a fuzzing mechanism that alters the states of (or creates new)
+ * input objects for methods under test to generate more object states and potentially trigger more
+ * branches, improving coverage for the program under test.
+ *
+ * <p>[TODO: As described in the GRT paper, "GRT Impurity" also generates more effective test with
+ * shorter length by reducing the number of redundant sequences that do not side-effect the state of
+ * an object. That is not implemented here.]
+ *
+ * <p>For details about fuzzing, see {@link #fuzz}.
+ */
+@SuppressWarnings("NotJavadoc") // perhaps https://github.com/google/error-prone/issues/3904
+public class GrtImpurity {
+
+  /** Ways to fuzz a string. */
+  private enum StringFuzzingOperation {
+    /** Insert a random character at a random index in the string. */
+    INSERT,
+
+    /** Remove a character at a random index in the string. */
+    REMOVE,
+
+    /** Replace a character at a random index in the string with a random character. */
+    REPLACE,
+
+    /** Take a random substring of the string. */
+    SUBSTRING
+  }
+
+  /** The standard deviation of the Gaussian distribution used to generate fuzzed numbers. */
+  private static final double GAUSSIAN_STD = GenInputsAbstract.grt_impurity_stddev;
+
+  /** Do not instantiate. */
+  private GrtImpurity() {
+    throw new Error("Do not instantiate");
+  }
+
+  /**
+   * Generate an extended fuzzed sequence for the given sequence.
+   *
+   * <p>This is the main entry point for GRT Impurity. This method appends additional fuzzing
+   * statements to the input sequence to fuzz the input object. This will generate more object
+   * states and potentially improve coverage for the unit test generated by Randoop by triggering
+   * more branches.
+   *
+   * <p>Different types have different fuzzing strategies:
+   *
+   * <ul>
+   *   <li><b>Numbers (int, short, long, float, double, and their wrapper classes):</b> Fuzzed using
+   *       a Gaussian distribution with the original value as the mean (mu). The fuzzed value is
+   *       generated as mu + N(0, GAUSSIAN_STD). For short values, if the fuzzed value exceeds the
+   *       short range, it is truncated.
```

Comment:
I don't see code for that truncation in this pull request.  (A narrowing conversion just throws away the higher-order bits.)
Also, why do other numbers overflow but this one is truncated?  That is inconsistent.

Reply:
Truncation (I should use a better word) is done automatically by Integer.shortValue() when fuzzing shorts. I should have mentioned this explicitly. I assumed that fuzzing a short outside its range would be significantly more common so I mentioned it, but I agree that I need to be consistent in my explanation.

What do you recommend? Should I mention that overflow/underflow can occur, which contradicts the Gaussian distribution for all primitive number types? Or do you think it's unnecessary to mention this at all?

Reply:
It's not "truncation", but modulus or bit masking.
I don't think it is necessary to mention in a Javadoc comment.

Reply:
Thank you for the suggestions!